CombinatorialGames/Game/Canonical.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2025 Tristan Figueroa-Reid. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tristan Figueroa-Reid
+-/
+
+import CombinatorialGames.Game.Birthday
+
+/-!
+# Canonical games
+
+For any game G, its canonical game G' is the unique IGame game with
+smallest birthday such that G'.Fits G.
+From the literature, this file provides an explicit (though noncomputable) construction of canonical
+games through undominating and unreversing games.
+
+## Todo
+- Define (un)reversibility
+-/
+
+universe u
+
+namespace IGame
+
+/-- Undominating a game. This returns garbage values on non-short games -/
+noncomputable def undominate (x : IGame) : IGame :=
+  {{y ∈ Set.range fun z : x.leftMoves ↦ undominate z | ∀ z ∈ x.leftMoves, ¬y < z} |
+    {y ∈ Set.range fun z : x.rightMoves ↦ undominate z | ∀ z ∈ x.rightMoves, ¬z < y}}ᴵ
+termination_by x
+decreasing_by igame_wf
+
+theorem birthday_undominate_le (x : IGame) : x.undominate.birthday ≤ x.birthday := by
+  rw [undominate, birthday_le_iff']
+  have (w) (hw : IsOption w x) := (birthday_undominate_le w).trans_lt (birthday_lt_of_isOption hw)
+  aesop
+termination_by x
+decreasing_by igame_wf
+
+theorem undominate_def {x : IGame} : x.undominate =
+    {{y ∈ undominate '' x.leftMoves | ∀ z ∈ x.leftMoves, ¬y < z} |
+      {y ∈ undominate '' x.rightMoves | ∀ z ∈ x.rightMoves, ¬z < y}}ᴵ := by
+  rw [undominate]
+  simp
+
+@[simp]
+theorem leftMoves_undominate {x : IGame} :
+    x.undominate.leftMoves = {y ∈ undominate '' x.leftMoves | ∀ z ∈ x.leftMoves, ¬y < z} := by
+  rw [undominate_def]
+  exact leftMoves_ofSets ..
+
+@[simp]
+theorem rightMoves_undominate {x : IGame} :
+    x.undominate.rightMoves = {y ∈ undominate '' x.rightMoves | ∀ z ∈ x.rightMoves, ¬z < y} := by
+  rw [undominate_def]
+  exact rightMoves_ofSets ..
+
+instance {x : IGame} [hx : Short x] : Short (undominate x) := by
+  rw [short_iff_birthday_finite] at hx ⊢
+  exact (birthday_undominate_le x).trans_lt hx
+
+theorem _root_.Set.neg_setOf {α : Type*} [InvolutiveNeg α] (p : α → Prop) :
+    -{x | p x} = {x | p (-x)} :=
+  rfl
+
+theorem _root_.Set.image_neg_of_apply_neg {α β : Type*} [InvolutiveNeg α] [InvolutiveNeg β]
+    {s : Set α} {f : α → β} (H : ∀ x ∈ s, f (-x) = -f x) : f '' (-s) = -f '' s := by
+  ext
+  rw [Set.mem_image, ← (Equiv.neg _).exists_congr_right]
+  aesop
```

Comment:
I'll PR these to Mathlib tomorrow, in the meanwhile we might want to separate them and put them near the beginning of the file.